### PR TITLE
Fix tensorflow 2.5.2 cross test failure

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

Fix tensorflow 2.5.2 cross test failure
Tensorflow release 2.5.2 today https://pypi.org/project/tensorflow/2.5.2/ , and it cause this failure happen.

## How is this patch tested?

Unit tests.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 1]
```diff
@@ -912,7 +912,7 @@ def test_import_tensorflow_with_fluent_autolog_enables_tf_autologging():
 
     # NB: For backwards compatibility, fluent autologging enables TensorFlow and
     # Keras autologging upon tensorflow import in TensorFlow 2.5.1
```
Comment (liangz1) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:914]: The comments are also outdated here.

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 2]
```diff
@@ -927,7 +927,7 @@ def test_import_tf_keras_with_fluent_autolog_enables_tf_autologging():
 
     # NB: For backwards compatibility, fluent autologging enables TensorFlow and
     # Keras autologging upon tf.keras import in TensorFlow 2.5.1
```
Comment (liangz1) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:929]: ditto

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 3]
```diff
@@ -912,7 +912,7 @@ def test_import_tensorflow_with_fluent_autolog_enables_tf_autologging():
 
     # NB: For backwards compatibility, fluent autologging enables TensorFlow and
     # Keras autologging upon tensorflow import in TensorFlow 2.5.1
-    if Version(tf.__version__) != Version("2.5.1"):
+    if Version(tf.__version__) >= Version("2.6"):
```
Comment (liangz1) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:915]: Why do we also want to disable `Version(tf.__version__) < Version("2.5.1")`? it looks like only 2.5.1 and 2.5.2 are causing failures.

Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py:915]: @dbczumar 
I think previous code logic here is wrong. Only >= 2.6 version will redirect keras autologging to tensorflow autologging.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 ! Can we update the comments in the tests as well, as @liangz1 pointed out?

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py (hunk 3 above)
Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py:915]: CC @liangz1 

For more context, you can read this PR:
https://github.com/mlflow/mlflow/pull/4766

Simple explanation:
In Tensorflow >= 2.6, the Tensorflow embedded keras will be linked to the solely installed keras (in contrast, previous Tensorflow embedded keras was a different module with the solely installed keras)

The change in Tensorflow >= 2.6, cause the mlflow.tensorflow.autolog (including patching on tf and embeded tf.keras) and mlflow.keras.autolog (patching on keras ) conflicts. So, as a workaround, for TF >= 2.6, we disable mlflow.keras.autolog, but instead, we only trigger mlflow.tensorflow.autolog

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 4]
```diff
@@ -925,9 +925,9 @@ def test_import_tf_keras_with_fluent_autolog_enables_tf_autologging():
 
     assert not autologging_is_disabled(mlflow.tensorflow.FLAVOR_NAME)
 
-    # NB: For backwards compatibility, fluent autologging enables TensorFlow and
-    # Keras autologging upon tf.keras import in TensorFlow 2.5.1
-    if Version(tf.__version__) != Version("2.5.1"):
+    # NB: In Tensorflow >= 2.6, we redirect keras autologging to tensorflow autologging
+    # so the original keras autologging is disabled
+    if Version(tf.__version__) >= Version("2.6"):
         assert autologging_is_disabled(mlflow.keras.FLAVOR_NAME)
```
Comment (harupy) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:931]: In `tensorflow < 2.6`, does `import tensorflow.keras` enable keras autologging?

Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py]: No.

Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py]: Good catch. We haven't import keras in the test, add `import keras` to make testing meaningful.

Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py]: In tensorflow < 2.6, keras is not a dependency of tensorflow, in the cross test we cannot import keras module

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 5]
```diff
@@ -925,9 +926,10 @@ def test_import_tf_keras_with_fluent_autolog_enables_tf_autologging():
 
     assert not autologging_is_disabled(mlflow.tensorflow.FLAVOR_NAME)
 
-    # NB: For backwards compatibility, fluent autologging enables TensorFlow and
-    # Keras autologging upon tf.keras import in TensorFlow 2.5.1
-    if Version(tf.__version__) != Version("2.5.1"):
+    # NB: In Tensorflow >= 2.6, we redirect keras autologging to tensorflow autologging
+    # so the original keras autologging is disabled
+    if Version(tf.__version__) >= Version("2.6"):
+        import keras  # pylint: disable=unused-variable,unused-import
```
Comment (WeichenXu123) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:932]: Add `import keras` to make test following test `autologging_is_disabled(mlflow.keras.FLAVOR_NAME)` meaningful

Comment (harupy) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py:932]: Does `import tensorflow.keras` run `import keras` or not?

Comment (harupy) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py:932]: I'd prefer to create a separate test (e.g. `test_import_keras_with_fluent_autolog_does_not_enable_keras_autologging`)

Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py]: > Does import tensorflow.keras run import keras or not?

For TF >= 2.6, Yes.
Remove `import keras` under this case.

**Comment by WeichenXu123** [comment]
Seems a new error (irrelative to this PR) happen:
```
E   tensorflow.python.framework.errors_impl.AlreadyExistsError: Another metric with the same name already exists.
```
This happen under TF 2.6 installed (it will install a incompatible keras version (keras 2.7) as dependency)

**Comment by jwyyy** [comment]
> Seems a new error (irrelative to this PR) happen:
> 
> ```
> 
> E   tensorflow.python.framework.errors_impl.AlreadyExistsError: Another metric with the same name already exists.
> 
> ```
> 
> This happen under TF 2.6 installed (it will install a incompatible keras version (keras 2.7) as dependency)

This has caused flavor check failure in other recent PRs as well.

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py (hunk 5 above)
Comment (WeichenXu123) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py]: > I'd prefer to create a separate test (e.g. test_import_keras_with_fluent_autolog_does_not_enable_keras_autologging)

We already has it. "test_import_keras_with_fluent_autolog_enables_tensorflow_autologging"

The test `test_import_tensorflow_with_fluent_autolog_enables_tf_autologging` is a bit different, it want to test we import tensorflow first, then import keras first, different order leads to different code path coverage.

Comment (harupy) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py]: > We already has it. "test_import_keras_with_fluent_autolog_enables_tensorflow_autologging"

Ah I missed that.

**Comment by harupy** [comment]
https://github.com/tensorflow/tensorflow/issues/52922#issuecomment-960418731 says:

**Comment by BenWilson2** [comment]
TF 2.6.2 released to pypi about 2 hours ago with a version fix in dependencies.

**Pull request merged**